File: packages/java/src/main/java/com/tesote/sdk/internal/InMemoryCacheBackend.java
package com.tesote.sdk.internal;

import com.tesote.sdk.CacheBackend;

import java.time.Duration;
import java.time.Instant;
import java.util.Map;
import java.util.Optional;
import java.util.concurrent.ConcurrentHashMap;

/**
 * Default thread-safe in-memory backend with TTL expiry. Not LRU-bounded — the
 * design doc suggests LRU but the default is fine for the bootstrap; users with
 * real cache pressure plug in their own backend.
 */
public final class InMemoryCacheBackend implements CacheBackend {
    private record Entry(byte[] value, Instant expiresAt) {}

    private final Map<String, Entry> store = new ConcurrentHashMap<>();

    @Override
    public Optional<byte[]> get(String key) {
        Entry e = store.get(key);
        if (e == null) return Optional.empty();
        if (Instant.now().isAfter(e.expiresAt())) {
            store.remove(key, e);
            return Optional.empty();
        }
        return Optional.of(e.value());
    }

    @Override
    public void put(String key, byte[] value, Duration ttl) {
        store.put(key, new Entry(value, Instant.now().plus(ttl)));
    }

    @Override
    public void invalidate(String key) {
        store.remove(key);
    }

    @Override
    public void invalidatePrefix(String prefix) {
        store.keySet().removeIf(k -> k.startsWith(prefix));
    }
}

File: packages/csharp/src/Internal/InMemoryCacheBackend.cs
using System;
using System.Collections.Concurrent;

namespace Tesote.Sdk.Internal;

/// <summary>
/// Default thread-safe in-memory backend with TTL expiry. Not LRU-bounded —
/// users with real cache pressure plug in their own backend.
/// </summary>
public sealed class InMemoryCacheBackend : ICacheBackend
{
    private readonly ConcurrentDictionary<string, Entry> _store = new();

    /// <inheritdoc />
    public byte[]? Get(string key)
    {
        ArgumentNullException.ThrowIfNull(key);
        if (!_store.TryGetValue(key, out var entry))
        {
            return null;
        }
        if (DateTimeOffset.UtcNow > entry.ExpiresAt)
        {
            _store.TryRemove(key, out _);
            return null;
        }
        return entry.Value;
    }

    /// <inheritdoc />
    public void Put(string key, byte[] value, TimeSpan ttl)
    {
        ArgumentNullException.ThrowIfNull(key);
        ArgumentNullException.ThrowIfNull(value);
        _store[key] = new Entry(value, DateTimeOffset.UtcNow.Add(ttl));
    }

    /// <inheritdoc />
    public void Invalidate(string key)
    {
        ArgumentNullException.ThrowIfNull(key);
        _store.TryRemove(key, out _);
    }

    /// <inheritdoc />
    public void InvalidatePrefix(string prefix)
    {
        ArgumentNullException.ThrowIfNull(prefix);
        foreach (var k in _store.Keys)
        {
            if (k.StartsWith(prefix, StringComparison.Ordinal))
            {
                _store.TryRemove(k, out _);
            }
        }
    }

    private readonly record struct Entry(byte[] Value, DateTimeOffset ExpiresAt);
}
